crates/ruff/src/checkers/ast.rs

```diff
@@ -3900,7 +3900,10 @@ impl<'a> Checker<'a> {
     }
 
     pub const fn execution_context(&self) -> ExecutionContext {
-        if self.in_type_checking_block || self.in_annotation {
+        if self.in_type_checking_block
+            || self.in_annotation
+            || self.in_deferred_string_type_definition
```

Comment:
We still want to omit `self.in_deferred_type_definition`, since if that's true but `self.in_annotation` is `False`, then we're in a runtime-required future type.